src/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs

```diff
@@ -24,6 +24,70 @@ private struct DateTimeParseData
             public byte OffsetToken;
         }
 
+        public static string FormatDateTimeOffset(DateTimeOffset value)
+        {
+            Span<byte> span = stackalloc byte[JsonConstants.MaximumFormatDateTimeOffsetLength];
+
+            JsonWriterHelper.WriteDateTimeOffsetTrimmed(span, value, out int bytesWritten);
+
+            return JsonReaderHelper.GetTextFromUtf8(span.Slice(0, bytesWritten));
+        }
+
+        public static string FormatDateTime(DateTime value)
+        {
+            Span<byte> span = stackalloc byte[JsonConstants.MaximumFormatDateTimeOffsetLength];
+
+            JsonWriterHelper.WriteDateTimeTrimmed(span, value, out int bytesWritten);
+
+            return JsonReaderHelper.GetTextFromUtf8(span.Slice(0, bytesWritten));
+        }
+
+        public static bool TryParseAsISO(ReadOnlySpan<char> source, out DateTime value)
+        {
+            if (!IsValidDateTimeOffsetParseLength(source.Length))
+            {
+                value = default;
+                return false;
+            }
+
+            Debug.Assert(JsonReaderHelper.GetUtf8ByteCount(source) <= JsonConstants.StackallocThreshold);
```

Comment:
What guarantees this?  What if the source is erroneously made up of MaximumEscapedDateTimeOffsetParseLength Unicode characters?

Reply:
The `IsValidDateTimeOffsetParseLength` check above (https://github.com/dotnet/corefx/pull/41940#discussion_r338400888):

> In fact, the above check using `IsValidDateTimeOffsetParseLength` guarantees its between 10 and 6*(33+16-7), i.e. 252.

`JsonConstants.StackallocThreshold` is 256 :smile:

Reply:
Ah, I see what you mean now. If the `char`s in the span expands to more UTF8 bytes?

Reply:
Right.

Reply:
Good point. We should add a test case for that to make sure subsequent changes don't fall into the same issue.

That said, instead of calling `JsonReaderHelper.GetUtf8ByteCount`, can we just assume the worst case, and do:
```C#
int length = source.Length * JsonConstants.MaxExpansionFactorWhileTranscoding;
```

In most cases, the length will still be less than 256, and we avoid the double read over the source span.

Reply:
Opened https://github.com/dotnet/corefx/pull/42574